File: frontend/src/types/property.ts
// Matches backend: app/models/property.py
export type PropertyType = 'apartment' | 'house' | 'studio' | 'shared'
export type PropertyStatus = 'available' | 'rented' | 'maintenance' | 'offline'

// Matches backend: app/schemas/property.py PropertyRead
export interface Property {
  id: number
  landlord_id: number
  property_manager_id?: number | null
  title: string
  description: string
  deposit_amount?: number
  service_fee_rate?: number | null
  address: string
  district: string
  price_monthly: number
  area_sqm: number | null
  bedrooms: number
  bathrooms: number
  property_type: PropertyType
  status: PropertyStatus
  latitude: number | null
  longitude: number | null
  created_at: string
  updated_at: string
  images?: PropertyImage[]
  primary_image_url?: string | null
  similarity?: number | null
}

// Matches backend: app/schemas/property.py PropertyCreate
export interface PropertyCreate {
  title: string
  description?: string
  address: string
  district: string
  price_monthly: number
  area_sqm?: number
  bedrooms?: number
  bathrooms?: number
  property_type?: PropertyType
  status?: PropertyStatus
  latitude?: number
  longitude?: number
  landlord_id: number
  property_manager_id?: number | null
}

// Matches backend: app/schemas/property.py PropertyUpdate
export interface PropertyUpdate {
  title?: string
  description?: string
  address?: string
  district?: string
  price_monthly?: number
  area_sqm?: number
  bedrooms?: number
  bathrooms?: number
  property_type?: PropertyType
  status?: PropertyStatus
  latitude?: number
  longitude?: number
  property_manager_id?: number | null
}

// Matches backend: app/schemas/property.py PropertySearchResult
export interface PropertySearchResult extends Property {
  similarity: number | null
}

// Search parameters matching backend query params (GET /properties/search)
// Note: country/overseas_area are frontend-only filters; backend maps them to district
export interface PropertySearchParams {
  q?: string              // Natural language query → backend pgvector search
  country?: string         // ⚠️ Frontend filter; mapped to district before sending
  district?: string        // Backend-supported
  overseas_area?: string   // ⚠️ Frontend filter; mapped to district before sending
  price_min?: number       // Backend-supported
  price_max?: number       // Backend-supported
  bedrooms?: number        // Backend-supported
  property_type?: PropertyType  // Backend-supported
  limit?: number           // Backend-supported (1-100)
}


// Property Image
export interface PropertyImage {
  id: number
  property_id: number
  filename: string
  original_name: string
  mime_type: string
  file_size: number
  sort_order: number
  is_primary: boolean
  created_at: string
}

File: backend/app/schemas/property.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.property_image import PropertyImageRead

from app.models.property import PropertyStatus, PropertyType


class PropertyBase(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    address: str = Field(min_length=1, max_length=300)
    district: str = Field(min_length=1, max_length=100)
    price_monthly: Decimal = Field(ge=0)
    area_sqm: Decimal | None = Field(default=None, gt=0)
    bedrooms: int = Field(default=0, ge=0)
    bathrooms: int = Field(default=0, ge=0)
    property_type: PropertyType = PropertyType.apartment
    status: PropertyStatus = PropertyStatus.available
    latitude: Decimal | None = Field(default=None, ge=-90, le=90)
    longitude: Decimal | None = Field(default=None, ge=-180, le=180)
    deposit_amount: int | None = None
    service_fee_rate: float | None = None

class PropertyCreate(PropertyBase):
    landlord_id: int
    property_manager_id: int | None = None
    repair_worker_id: int | None = None


class PropertyUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    address: str | None = Field(default=None, min_length=1, max_length=300)
    district: str | None = Field(default=None, min_length=1, max_length=100)
    price_monthly: Decimal | None = Field(default=None, ge=0)
    area_sqm: Decimal | None = Field(default=None, gt=0)
    bedrooms: int | None = Field(default=None, ge=0)
    bathrooms: int | None = Field(default=None, ge=0)
    property_type: PropertyType | None = None
    status: PropertyStatus | None = None
    latitude: Decimal | None = Field(default=None, ge=-90, le=90)
    longitude: Decimal | None = Field(default=None, ge=-180, le=180)
    property_manager_id: int | None = None
    repair_worker_id: int | None = None


class PropertyRead(PropertyBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    landlord_id: int
    property_manager_id: int | None = None
    repair_worker_id: int | None = None
    created_at: datetime
    updated_at: datetime
    images: list[PropertyImageRead] = []

    @property
    def primary_image_url(self) -> str | None:
        for img in self.images:
            if img.is_primary:
                return f"/api/v1/uploads/{img.filename}"
        return None



class PropertySearchResult(PropertyBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    landlord_id: int
    property_manager_id: int | None = None
    repair_worker_id: int | None = None
    created_at: datetime
    updated_at: datetime
    images: list[PropertyImageRead] = []
    similarity: float | None = None

    @property
    def primary_image_url(self) -> str | None:
        for img in self.images:
            if img.is_primary:
                return f"/api/v1/uploads/{img.filename}"
        return None
